uniffi_bindgen/src/interface/types/finder.rs

```diff
@@ -0,0 +1,176 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! # Helpers for finding the named types defined in a UDL interface.
+//!
+//! This module provides the [`TypeFinder`] trait, an abstraction for walking
+//! the weedle parse tree, looking for type definitions, and accumulating them
+//! in a [`TypeUniverse`].
+//!
+//! The type-finding process only discovers very basic information about names
+//! and their corresponding types. For example, it can discover that "Foobar"
+//! names a Record, but it won't discover anything about the fields of that
+//! record.
+//!
+//! Factoring this functionality out into a separate phase makes the subsequent
+//! work of more *detailed* parsing of the UDL a lot simpler, we know how to resolve
+//! names to types when building up the full interface definition.
+
+use std::convert::TryFrom;
+
+use anyhow::{bail, Result};
+
+use super::super::enum_::EnumAttributes;
+use super::{Type, TypeUniverse};
+
+/// Trait to help with an early "type discovery" phase when processing the UDL.
+///
+/// Ths trait does structural matching against weedle AST nodes from a parsed
+/// UDL file, looking for all the newly-defined types in the file and accumulating
+/// them in the given `TypeUniverse`.
+pub(in super::super) trait TypeFinder {
```

Comment:
There's a lot of `pub(super)` and a bit of `pub(in super:super)` going on here which is kinda gross, but I didn't want to make more invasive code changes that might have been necessary to clean up the willy-nilly trait and struct field access that was going on thanks to all these previously being in a single file...